lib/std/os/linux/io_uring.zig

```diff
@@ -0,0 +1,877 @@
+    /// Returns the number of SQEs submitted.
+    /// Matches the implementation of io_uring_submit() in liburing.
+    pub fn submit(self: *IO_Uring) !u32 {
+        return self.submit_and_wait(0);
+    }
+
+    /// Like submit(), but allows waiting for events as well.
+    /// Returns the number of SQEs submitted.
+    /// Matches the implementation of io_uring_submit_and_wait() in liburing.
+    pub fn submit_and_wait(self: *IO_Uring, wait_nr: u32) !u32 {
+        var submitted = self.flush_sq();
+        var flags: u32 = 0;
+        if (self.sq_ring_needs_enter(submitted, &flags) or wait_nr > 0) {
+            if (wait_nr > 0 or (self.flags & linux.IORING_SETUP_IOPOLL) != 0) {
+                flags |= linux.IORING_ENTER_GETEVENTS;
+            }
+            return try self.enter(submitted, wait_nr, flags);
+        }
+        return submitted;
+    }
+
+    /// Tell the kernel we have submitted SQEs and/or want to wait for CQEs.
+    /// Returns the number of SQEs submitted.
+    pub fn enter(self: *IO_Uring, to_submit: u32, min_complete: u32, flags: u32) !u32 {
+        assert(self.fd >= 0);
+        const res = linux.io_uring_enter(self.fd, to_submit, min_complete, flags, null);
+        switch (linux.getErrno(res)) {
+            0 => {},
+            // The kernel was unable to allocate memory or ran out of resources for the request.
+            // The application should wait for some completions and try again:
+            linux.EAGAIN => return error.SystemResources,
+            // The application attempted to overcommit the number of requests it can have pending.
+            // The application should wait for some completions and try again:
+            linux.EBUSY => return error.CompletionQueueOvercommitted,
+            // The SQE `fd` is invalid, or IOSQE_FIXED_FILE was set but no files were registered:
+            linux.EBADF => return error.FileDescriptorInvalid,
+            // The buffer is outside the process' accessible address space, or IORING_OP_READ_FIXED
+            // or IORING_OP_WRITE_FIXED was specified but no buffers were registered, or the range
+            // described by `addr` and `len` is not within the buffer registered at `buf_index`:
+            linux.EFAULT => return error.BufferInvalid,
+            // The SQE is invalid, or valid but the ring was setup with IORING_SETUP_IOPOLL:
+            linux.EINVAL => return error.SubmissionQueueEntryInvalid,
+            linux.ENXIO => return error.RingShuttingDown,
+            // The kernel believes our `self.fd` does not refer to an io_uring instance,
+            // or the opcode is valid but not supported by this kernel (more likely):
+            linux.EOPNOTSUPP => return error.OpcodeNotSupported,
+            // The operation was interrupted by a delivery of a signal before it could complete.
+            // This can happen while waiting for events with IORING_ENTER_GETEVENTS:
+            linux.EINTR => return error.SignalInterrupt,
```

Comment:
Yes I thought of that, but the kernel is free to overload errors as it has done, and then our unreachable would be undefined behavior rather than a safe error.

Reply:
that is correct; but in general the zig standard library assumes (with undefined behaviour) that the kernel doesn't overload errors.
You may want to open a new issue for that?

Reply:
Sure! https://github.com/ziglang/zig/issues/6389